src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.SingleEntry.targets

```diff
@@ -2,24 +2,24 @@
 
   <PropertyGroup Condition="'$(RuntimeIdentifier)' != '' and '$(NETCoreSdkPortableRuntimeIdentifier)' != ''">
     <!-- Define the name of the runtime specific compiler package to import -->
-    <_targetOSPkg>$(NETCoreSdkPortableRuntimeIdentifier.SubString(0, $(NETCoreSdkPortableRuntimeIdentifier.LastIndexOf('-'))))</_targetOSPkg>
-    <_indexOfHyphen>$(_targetOSPkg.IndexOf('-'))</_indexOfHyphen>
-    <_targetOS Condition="'$(_indexOfHyphen)' &gt; -1">$(_targetOSPkg.SubString(0, $(_indexOfHyphen)))</_targetOS>
-    <_targetOS Condition="'$(_indexOfHyphen)' == -1">$(_targetOSPkg)</_targetOS>
-    <_indexOfPeriod>$(_targetOS.IndexOf('.'))</_indexOfPeriod>
-    <_targetOS Condition="'$(_indexOfPeriod)' &gt; -1">$(_targetOS.SubString(0, $(_indexOfPeriod)))</_targetOS>
+    <_useTPI Condition="'$(TargetPlatformIdentifier)' != '' and '$(TargetPlatformIdentifier)' != 'unix'">true</_useTPI>
```

Comment:
`TargetPlatformIdentifier` is derived from TFM. It does not need to match the actual target RID. I think that the `unix` that you have special-cased here is just one of the cases where it does not work.

Can we always use `RuntimeIdentifier` to identify the target OS+arch?

Reply:
For example, one can create hello world, set `TargetFramework` in .csproj to `net8.0-windows` and publish for `-r osx-x64`. The SDK won't complain about it.

Reply:
`TargetPlatformIdentifier` convention is used by MAUI and Xamarin SDK for iOS and Android like platforms. In runtime libraries, we use it to differentiate between -windows and -unix partials. That's where the !unix condition is coming from (because tests were failing).

> Can we always use `RuntimeIdentifier` to identify the target OS+arch?

We can't use `RuntimeIdentifier` can be non-portable (win10-x64, osx.13.1-arm64 etc.), so `NETCoreSdkPortableRuntimeIdentifier` helps.

We can't use `NETCoreSdkPortableRuntimeIdentifier` everywhere because for `ios-arm64` MAUI app published from `osx-arm64` machine, `NETCoreSdkPortableRuntimeIdentifier` is `ios-arm64`.

Reply:
> We can't use RuntimeIdentifier because it can be non-portable (win10-x64, osx.13.1-arm64 etc.), so NETCoreSdkPortableRuntimeIdentifier helps resolving the portable platform RID.

I think we should make best effort with normalizing the non-portable identifiers into portable ones. E.g. normalize win10-x64 into win, etc.

The normalization is impossible to do reliably for custom Linux RIDs (e.g. ubuntu). We do not know what the user has or intends to do. We should make it possible to override it explicitly on the command line, so that people can always make things work.

> `NETCoreSdkPortableRuntimeIdentifier`

`NETCoreSdkPortableRuntimeIdentifier` is a host identifier. It is a poor indicator of the target. We should avoid making fragile assumptions about which host/target combinations work and which do not.

Reply:
> `NETCoreSdkPortableRuntimeIdentifier` is a host identifier. It is a poor indicator of the target. We should avoid making fragile assumptions about which host/target combinations work and which do not.

This property is used by the SDK for `--ucr, --use-current-runtime` option and auto-inferred RIDs (in .NET 6 onward). Its value is `linux-x64` on source-built / non-portable `ubuntu-x64`. Source build also relies on it for some cases (like crossgen2 package resolution etc.).

Reply:
Right, this property works well as the indicator of target when host and target are the same. 

We are trying to make this logic work for cases where host and target are not the same. It means that this property is not very useful for identifying the target.

Reply:
Yes, we are using `NETCoreSdkPortableRuntimeIdentifier` to identify host OS and in case of non-portable RID, using it to identify target OS.

Reply:
> in case of non-portable RID, using it to identify target OS.

How do you know that you have encountered non-portable target RID? Non-portable target RIDs are a guess work. I would avoid dealing with them and just let the people set it explicitly if they need them.

Reply:
I agree that `_useTPI` heuristic is fragile and it will scale better if we keep it simple. I have pushed a change which starts deriving `_targetOS` value from `RuntimeIdentifier` and handles a few known corner cases.